spec/filesystem.rb

```diff
@@ -38,9 +38,10 @@ class FileSystem
   # called on a `FileSystemConfigurator` instance and will then be correctly
   # invoked when `FileSystem.test` is run with that instance. If these were run
   # directly outside of a `test` block then the real file system would be used.
-  delegate :mkdir_p, :touch, to: FileUtils
+  delegate :mkdir_p, :touch, :rm_rf, to: FileUtils
   delegate :write, to: File
   delegate :dump, to: Metalware::Data
+  delegate :enable!, to: Metalware::Plugins
```

Comment:
I don't think `enable!` should be delegated to `Metalware::Plugins`. I would expect `FileSystem.enable!` to enable `FakeFS` as a whole and not enable the plugins. A more descriptive method would be `enable_plugin`.

Also I don't think this needs to be a `!` method unless an additional "safe" method is also specified. There are plenty of cases where ruby methods are "dangerous" without having the `!` (e.g. `array.delete`).

Reply:
Addressed your first point in https://github.com/alces-software/metalware/commit/a40c5d8448df93d53785929288e5f9af9abe10ad; and extracted these setup methods to their own module in https://github.com/alces-software/metalware/commit/db092b03bf7eaf99e61c89df8473e2746e83b612.

Regarding the second point, I'm not sure - I try to define `!` methods even if there isn't a corresponding 'safe' method, for any situation which will potentially unexpectedly: (a) effect things on disk, (b) mutate data, or (c) normally raise an exception. In this case this does both (a) and (b) so I find it helpful to know that I may need to be a bit careful when calling this method.

In general I find this works reasonably for my code as I try to minimize the places which will do these things and so the `!` methods stand out a bit, but it can be a bit fuzzy when using a `!` in a method will be helpful. I'm open to changing my opinion on this however :slightly_smiling_face:.

If we do decide to change this, this is an issue which applies more generally across Metalware, as we have various places where we define `!` methods without any corresponding `!`-less methods; for instance:

```
$ git grep -E 'def [a-z_]+!'
spec/filesystem.rb:  def debug!
src/cli.rb:    def run!
src/commander_extensions.rb:    def validate_syntax!
src/commander_extensions.rb:    def validate_correct_number_of_args!(args)
src/commands/hunter.rb:      def listen!
src/gui/bin/setup:def system!(*args)
src/gui/bin/update:def system!(*args)
src/hash_mergers/hash_merger.rb:      def replace_file_with_same_basename!(files_namespace, file_identifier)
src/plugins.rb:      def enable!(plugin_name)
src/plugins.rb:      def disable!(plugin_name)
src/plugins.rb:      def validate_plugin_exists!(plugin_name)
src/plugins.rb:      def update_enabled_plugins!(new_enabled_plugins)
src/plugins.rb:    def enable!
```

Reply:
Okay checking the ruby doc the convention is the `!` method is a dangerous method. However it also specifies that there is almost always a safe method for every `!`. I'm happy with it as is, metalware has been playing hard and fast with this convention anyway.

Reply:
OK great, thanks for pointing it out though; it's something to consider in future.